File: apps/system/api/domain/workspace/subdomain.go
package workspace

import (
	"fmt"
	domainError "github.com/ryo034/react-go-template/apps/system/api/domain/shared/error"
	"github.com/ryo034/react-go-template/apps/system/api/domain/shared/validation"
	"regexp"
	"strings"
)

type Subdomain struct {
	v string
}

const (
	InvalidWorkspaceSubdomain domainError.MessageKey = "invalid.workspace.subdomain"
)

const maxLength = 63
const minLength = 1

func isValidSubdomain(subdomain string) (bool, error) {
	if len(subdomain) < minLength || len(subdomain) > maxLength {
		return false, nil
	}
	return regexp.MatchString(`^[a-z0-9]+(-[a-z0-9]+)*$`, subdomain)
}

func NewSubdomain(v string) (Subdomain, error) {
	errs := validation.NewErrors()
	trimmed := strings.TrimSpace(v)
	ok, err := isValidSubdomain(trimmed)
	if err != nil || !ok {
		errs.Append(InvalidWorkspaceSubdomain, nil, fmt.Sprintf("invalid workspace domain: %s", v))
		return Subdomain{}, errs
	}
	return Subdomain{trimmed}, nil
}

func (d Subdomain) ToString() string {
	return d.v
}

File: apps/system/client/src/domain/workspace/subdomain.ts
import { Result } from "true-myth"
import { DomainError, ValueObject, domainKeys } from "~/domain/shared"

export class WorkspaceSubdomain extends ValueObject<string> {
  static max = 63
  static min = 1
  static pattern = /^[a-z0-9]+(-[a-z0-9]+)*$/

  private static isValidSubdomain(subdomain: string): boolean {
    return (
      WorkspaceSubdomain.pattern.test(subdomain) &&
      subdomain.length >= WorkspaceSubdomain.min &&
      subdomain.length <= WorkspaceSubdomain.max
    )
  }

  static create(v: string): Result<WorkspaceSubdomain, Error> {
    if (!WorkspaceSubdomain.isValidSubdomain(v)) {
      return Result.err(
        new DomainError({
          domainKey: domainKeys.WorkspaceSubdomain,
          value: v,
          message: `Invalid workspace domain: ${v}`
        })
      )
    }
    return Result.ok(new WorkspaceSubdomain(v))
  }

  toString(): string {
    return this.value
  }
}
